From 2e364a00be889cf642ba6f8748adcab7e143f7f4 Mon Sep 17 00:00:00 2001
From: Takashi Yano <takashi.yano@nifty.ne.jp>
Date: Mon, 18 Apr 2022 19:56:14 +0900
Subject: [PATCH 060/N] Cygwin: pty: Fix deadlock caused by Ctrl-C in
 setup_pseudoconsole().

- If Ctrl-C is pressed just after setup_pseudoconsole() is called,
  mintty stops to respond a while when CPU load is high. This patch
  fixes the issue.

Addresses: https://cygwin.com/pipermail/cygwin/2022-April/251272.html
Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
---
 winsup/cygwin/spawn.cc | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/winsup/cygwin/spawn.cc b/winsup/cygwin/spawn.cc
index 4004571..c9e1fb6 100644
--- a/winsup/cygwin/spawn.cc
+++ b/winsup/cygwin/spawn.cc
@@ -679,8 +679,12 @@ child_info_spawn::worker (const char *prog_arg, const char *const *argv,
 	  HANDLE h_stdin = handle (fileno_stdin, false);
 	  if (h_stdin == ptys_primary->get_handle_nat ())
 	    stdin_is_ptys = true;
+	  if (reset_sendsig)
+	    myself->sendsig = myself->exec_sendsig;
 	  ptys_primary->setup_for_non_cygwin_app (nopcon, envblock,
 						  stdin_is_ptys);
+	  if (reset_sendsig)
+	    myself->sendsig = NULL;
 	  ptys_primary->get_duplicated_handle_set (&ptys_handle_set);
 	  ptys_ttyp = (tty *) ptys_primary->tc ();
 	  ptys_need_cleanup = true;
-- 
2.9.0

